src/libxively/sft/xi_sft_logic.c

```diff
@@ -62,37 +68,178 @@ xi_state_t xi_sft_on_connected( xi_sft_context_t* context )
+                            context->update_current_file->size_in_bytes );
+
+                /* HERE process the bytes in sft_message_in->file_chunk.chunk */
+
+                if ( all_downloaded_bytes < context->update_current_file->size_in_bytes )
+                {
+                    /* current file is not downloaded yet, continue the download */
+
+                    xi_control_message_t* message_file_get_chunk =
+                        xi_control_message_create_file_get_chunk(
+                            context->update_current_file->name,
+                            context->update_current_file->revision, all_downloaded_bytes,
+                            XI_MIN( XI_SFT_DOWNLOAD_BYTES_PER_FILE_CHUNK,
+                                    context->update_current_file->size_in_bytes -
+                                        all_downloaded_bytes /* == bytes left */ ) );
+
+                    ( *context->fn_send_message )( context->send_message_user_data,
+                                                   message_file_get_chunk );
+                }
+                else
+                {
+                    /* file downloaded, continue with next file in list */
+
+                    {
+                        xi_control_message_t* message_file_status =
+                            xi_control_message_create_file_status(
+                                context->update_current_file->name,
+                                context->update_current_file->revision,
+                                XI_CONTROL_MESSAGE__SFT_FILE_STATUS_PHASE_DOWNLOADED,
+                                XI_CONTROL_MESSAGE__SFT_FILE_STATUS_CODE_SUCCESS );
+
+                        ( *context->fn_send_message )( context->send_message_user_data,
+                                                       message_file_status );
+
+                        message_file_status = xi_control_message_create_file_status(
+                            context->update_current_file->name,
+                            context->update_current_file->revision,
+                            XI_CONTROL_MESSAGE__SFT_FILE_STATUS_PHASE_PROCESSING,
+                            XI_CONTROL_MESSAGE__SFT_FILE_STATUS_CODE_SUCCESS );
+
+                        ( *context->fn_send_message )( context->send_message_user_data,
+                                                       message_file_status );
+
+                        message_file_status = xi_control_message_create_file_status(
+                            context->update_current_file->name,
+                            context->update_current_file->revision,
+                            XI_CONTROL_MESSAGE__SFT_FILE_STATUS_PHASE_FINISHED,
+                            XI_CONTROL_MESSAGE__SFT_FILE_STATUS_CODE_SUCCESS );
+
+                        ( *context->fn_send_message )( context->send_message_user_data,
+                                                       message_file_status );
```

Comment:
Unfortunately the Fingerprint check will probably be in the same `tick` as the DOWNLOADED, so it will still be confused on the backend. It's not our problem in the sense that we cannot fix it here, but we should definitely bubble up the fact that these might come out of order if the same backend system that  spits logging messages into new tasks is used for the SFT implementation.  I will bring it up with Tamas.